File: NetFramework/XHtmlEditor/VB/RichTextEditor/Dialogs/WordCount.vb
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms

Namespace RichTextEditor
	Public Partial Class WordCount
		Inherits Form
		Public Sub New()
			InitializeComponent()
		End Sub
		Public Sub New(text As String)
			InitializeComponent()

			Dim chars As Integer = 0
			Dim nonSpaceChars As Integer = 0
			Dim words As Integer = 0
			Dim paragraphs As Integer = 0
			Dim sentences As Integer = 0

			Dim length As Integer = text.Length

			' calculate statistics
			For i As Integer = 0 To length - 1
				Dim c As Char = text(i)

				' count chars
				chars += 1

				' count non-space chars
				If Not Char.IsWhiteSpace(c) Then
					nonSpaceChars += 1
				End If

				' count paragraphs
				If c = ControlChars.Lf OrElse i = length - 1 Then
					If i = length - 1 OrElse text(i + 1) <> ControlChars.Lf Then
						paragraphs += 1
					End If
				End If

				' count sentences
				If c = "."C OrElse c = "!"C OrElse c = "?"C OrElse i = length - 1 Then
					If i = length - 1 OrElse Char.IsWhiteSpace(text, i + 1) Then
						sentences += 1
					End If
				End If

				' count words
				If Char.IsLetterOrDigit(c) Then
					If i = length - 1 OrElse Not Char.IsLetterOrDigit(text, i + 1) Then
						words += 1
					End If
				End If
			Next

			' show statistics
			_lblWords.Text = words.ToString("n0")
			_lblChars.Text = chars.ToString("n0")
			_lblCharsNoSpace.Text = nonSpaceChars.ToString("n0")
			_lblParagraphs.Text = paragraphs.ToString("n0")
			_lblSentences.Text = sentences.ToString("n0")
            _lblWordLength.Text = String.Format("{0:n1}", Util.IIFF(words > 0, nonSpaceChars / CSng(words), 0.0F))
            _lblSentenceLength.Text = String.Format("{0:n1}", Util.IIFF(sentences > 0, nonSpaceChars / CSng(sentences), 0.0F))
            _lblParagraphLength.Text = String.Format("{0:n1}", Util.IIFF(paragraphs > 0, nonSpaceChars / CSng(paragraphs), 0.0F))
		End Sub
	End Class
End Namespace

File: NetFramework/XHtmlEditor/CS/RichTextEditor/Dialogs/WordCount.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;

namespace RichTextEditor
{
    public partial class WordCount : Form
    {
        public WordCount()
        {
            InitializeComponent();
        }
        public WordCount(string text)
        {
            InitializeComponent();

            int chars = 0;
            int nonSpaceChars = 0;
            int words = 0;
            int paragraphs = 0;
            int sentences = 0;

            int length = text.Length;

            // calculate statistics
            for (int i = 0; i < length; i++)
            {
                char c = text[i];

                // count chars
                chars++;

                // count non-space chars
                if (!char.IsWhiteSpace(c))
                {
                    nonSpaceChars++;
                }

                // count paragraphs
                if (c == '\n' || i == length - 1)
                {
                    if (i == length - 1 || text[i + 1] != '\n')
                    {
                        paragraphs++;
                    }
                }

                // count sentences
                if (c == '.' || c == '!' || c == '?' || i == length - 1)
                {
                    if (i == length - 1 || char.IsWhiteSpace(text, i + 1))
                    {
                        sentences++;
                    }
                }

                // count words
                if (char.IsLetterOrDigit(c))
                {
                    if (i == length - 1 || !char.IsLetterOrDigit(text, i + 1))
                    {
                        words++;
                    }
                }
            }

            // show statistics
            _lblWords.Text = words.ToString("n0");
            _lblChars.Text = chars.ToString("n0");
            _lblCharsNoSpace.Text = nonSpaceChars.ToString("n0");
            _lblParagraphs.Text = paragraphs.ToString("n0");
            _lblSentences.Text = sentences.ToString("n0");
            _lblWordLength.Text = string.Format("{0:n1}", words > 0 ? nonSpaceChars / (float)words : 0f);
            _lblSentenceLength.Text = string.Format("{0:n1}", sentences > 0 ? nonSpaceChars / (float)sentences : 0f);
            _lblParagraphLength.Text = string.Format("{0:n1}", paragraphs > 0 ? nonSpaceChars / (float)paragraphs : 0f);
        }
    }
}
